src/lib/y2partitioner/widgets/mkfs_options.rb

```diff
@@ -0,0 +1,161 @@
+require "yast"
+require "cwm"
+require "y2storage"
+require "y2partitioner/widgets/mkfs_optiondata"
+
+module Y2Partitioner
+  # Partitioner widgets
+  module Widgets
+    include Yast::Logger
+
+    # a module
+    module MkfsCommon
+      def initialize(controller, opts)
+        @controller = controller
+        @opts = opts
+        self.widget_id = "#{self.class}_#{object_id}"
+      end
+
+      def label
+        @opts[:label]
+      end
+
+      def help
+        "<p><b>#{@opts[:label].delete("&")}</b><br/>#{@opts[:help]}</p>"
+      end
+
+      def init
+        self.value = MkfsOptions.get_option(@controller.filesystem, @opts)
+      end
+
+      def store
+        MkfsOptions.set_option(@controller.filesystem, @opts, value)
+      end
+
+      def validate
+        return true unless @opts[:validate] && @opts[:error]
+
+        if !@opts[:validate][value]
+          Yast::Popup.Error(format(@opts[:error], value))
+          false
+        else
+          true
+        end
+      end
+    end
+
+    # a class
+    class MkfsOptions < CWM::CustomWidget
+      def initialize(controller)
+        textdomain "storage"
+        @controller = controller
+        self.handle_all_events = true
+      end
+
+      def help
+        Yast::CWM.widgets_in_contents(contents).find_all do |w|
+          w.respond_to?(:help)
+        end.map(&:help).join("\n")
+      end
+
+      def opt
+        [:notify]
+      end
+
+      def handle(event)
+        case event["ID"]
+        when :help
+          Yast::Wizard.ShowHelp(help)
+        end
+      end
+
+      def contents
+        # FIXME: add some VSpacing(1)?
+        # ???: contents is called 3 times for each dialog, so cache it
```

Comment:
Yes, the object oriented machinery on top of CWM does unexpected stuff in several cases. I also had to introduce some caches for the same reason... some of them way more weird than simply `@contents ||= `. :-)